pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -489,21 +503,59 @@ func (bfs *Service) IsActive() bool {
 	return bfs.isActive.Load()
 }
 
+// retry function with exponential backoff.
+func (bfs *Service) retry(action func() error) error {
+	var (
+		err     error
+		backoff = initialBackoff
+		timer   = time.NewTimer(0)
+	)
+	defer timer.Stop()
+	for range maxRetries {
+		if err = action(); err == nil {
+			return nil
+		}
+		if !timer.Stop() {
+			<-timer.C
+		}
+		timer.Reset(backoff)
+
+		select {
+		case <-timer.C:
+		case <-bfs.ctx.Done():
+			return bfs.ctx.Err()
+		}
```

Comment:
And BTW, when we're at the last iteration of the cycle, then we don't need to reset the timer and wait for a subsequent timer tick, because `action` won't be executed one more time.